### DISCUSSION TITLE
Feedback request: how can we make pyinfra’s documentation better

### DISCUSSION THREAD

**Discussion by Fizzadar** [pyinfra-dev/pyinfra#1168] [Ideas]
I’d like to improve pyinfra’s documentation, I think there’s a few things that open come up and clearly need documenting better. Off the top of my head some of these are:

- operation execution (when do things execute)
- how to get output from an operation
- how to read system data before operations execute (ie the `server.Command` fact)

Would be amazing to hear what other things are unclear to people, and any ideas on how to better document or structure the docs would be much appreciated! 🙏

**Comment by goetzk** [comment]
Some docs related comments from me:

Developing and using inventory and execution connectors is not obvious to me ; i've started documentation for that at https://github.com/pyinfra-dev/pyinfra/compare/3.x...medeopolis:pyinfra:3.x-documentation-updates?expand=1 and will share it once i've confirmed its correct.

Outstanding questions which i haven't added to that doc (because i don't have the answers yet)
- if a inventory connector needs configuration (to access an API), how should the configuration be 'inserted'. environment? config file? to be confirmed...
- how to use pyinfra logger appropriately , eg logging at -v , -vv, -vvv (this might be obvious from the code but i haven't checked the code yet)
- how to package functions (connector, fact gathering, operations) for use with pyinfra (including recommended tools, configuration which needs to be created, paths to observe, etc)

Edit: on the subject of developing connectors:
There are lots of helpers/formatters available in pyinfra that are used during connector development. its possible to cargo cult most of the process but it would still be good to have more """python comments""" explaining what things do.

**Reply by Fizzadar** [reply]
These are great suggestions, most of which are now in the refreshed docs as of 3.9!

**Comment by vram0gh2** [comment]
Hi - Firstly, I must thank you and all the contributors to the project.  Out of the various configuration automation systems out there, this fills an important role.  It's approachable, uses an actual scripting language for specifying what is to be done, only executes shell commands on the remote systems (agentless), and puts idempotency and declarative end state up-front.  Thank you!

I just started with pyinfra; for me, below are some missing, confusing, and ambiguous pieces in the docs.  If you are familiar with other config automation systems, you figure most of this out, but it shouldn't be expected or required.  **Warning:** this is a long post.

# Top-level
  The homepage mentions a lot of positive traits/differentiators pyinfra possesses, but to see what that really means, you have to go all over the docs.  Two ideas for improvement:
  * The top level index section of the docs could have a subsection mimicking the homepage, but link to the places in the docs that talk about those characteristics.
  * Use those traits as an explicit checklist of items the docs should discuss, ideally with consistent phrasing as with the homepage.
# [Getting Started](https://docs.pyinfra.com/en/3.x/getting-started.html)
  * Explicitly say whether the naming or quantity of host lists matter.  What does pyinfra do in terms of iterating over multiple variables, how are those names used?  What happens if a host is in two lists at the same time, does that matter?
  *  [Create a Deploy](https://docs.pyinfra.com/en/3.x/getting-started.html#create-a-deploy) raises many questions, but then says "That’s the basics of pyinfra!".  It feels glib and untrue.  You don't have to answer the following right there, but at least link to where these are discussed in the detailed chapters, and be sure to address them in those chapters.  Perhaps take the mini-index there and say for each of those index links, where the following are discussed:
     * Do the filenames of the CLI arguments matter or is it purely positional?
     * What are the requirements on the remote system for the command to be executed?
     * What identity is used to login to the system and perform any commands?  Will you get prompted for a password on the remote system if you don't use a key-based login?  If some commands can only be executed as `root` but the login identity isn't `root`, does pyinfra identify this beforehand and do something automagically (_e.g._, use `sudo` or `doas`) or alert the user?
       * Probably a good place for any admonishments about directly logging in as `root`.
     * The code block has a comment, `# Define some state - this operation will do nothing on subsequent runs`.  Why will it do nothing on subsequent runs?  Is there a cache of host state on the system running pyinfra?  Maybe change this to something more descriptive, like `# Define end state - checked each run, action taken only if not satisfied`?  Or something else more concise, but comprehensive and clear?
# [Using Operations](https://docs.pyinfra.com/en/3.x/using-operations.html)
  * The "What are operations?" block is confusing/ambiguous.  Maybe just explicitly say up-front that operations can either be a) imperative actions to be taken, or b) declaratively identify an end state to be achieved, in which case pyinfra figures out if anything needs to be done to reach that end state, and if so, does it.
  * It can be confusing when the `name` argument in the examples begin with imperative verbs, which may suggest it will attempt to do this each time it's run.  But because these operations are idempotent, that's not the case.  Say what the `name` argument is used for?
  * Early on, perhaps right after the first example, show how this gets turned into shell commands to be executed on the host?
  * Before getting into the [The host Object](https://docs.pyinfra.com/en/3.x/using-operations.html#the-host-object), perhaps discuss how the operations check the current state and only act if needed.  This can reinforce that things do get executed on the host to check the state and there can be an example case of what happens when the state matches the operation end state and when it doesn't.  In the rest of the page, what it means to "execute an operation" for a declarative operation is unclear.  Is it an atomic check of state and immediate action if the target end state is not met?  Does it follow a phased approach where first host state is checked, and in a separate phase all actions to be taken are executed?  If the latter, does the user know after the change detection and after and after any host state mutation?
  * [Host Facts](https://docs.pyinfra.com/en/3.x/using-operations.html#host-facts): it may help to remind the user that for individual declarative end state operations, the user doesn't have to check the fact associated with the current state themselves to determine if action is required for that operation.
  * [Host Facts](https://docs.pyinfra.com/en/3.x/using-operations.html#host-facts): there is an admonishment to only use immutable facts in deploy code.  Spell out what immutable means in this context?
  * [Fact Errors](https://docs.pyinfra.com/en/3.x/using-operations.html#fact-errors): it is mentioned "the host will be marked as failed just as it would when an operation fails".  But we haven't yet talked about what happens when an operation fails, how would the user already understand that?
  * [Operation Changes & Output](https://docs.pyinfra.com/en/3.x/using-operations.html#operation-changes-output): The first sentence is confusing without understanding pyinfra's execution model.  At least link to that section first?  When it is said that "all operations return ..." that implies the operation is complete, but then the sentence says "changes the operation will execute."  This sounds like an intuitive contradiction as written.
  * Is `.did_change` the only signal field these operations return?  It's just used in the example without discussing it first.
  * [The config Object](https://docs.pyinfra.com/en/3.x/using-operations.html#the-config-object): When should the `config` object be modified vs. using global variables?
  * When examining the host state to see if modifications are needed, what happens if the desired end state is the compound effect of multiple operations?  I am thinking about files for example, where you might upload a file and then modify it in a subsequent operation.
# [Inventory & Data](https://docs.pyinfra.com/en/3.x/inventory-data.html)
  * [Data Hierarchy](https://docs.pyinfra.com/en/3.x/inventory-data.html#data-hierarchy): mentions `all.py` - where is that discussed before?  Is this a special file that is read outside of the two arguments provided to the CLI?
# [Global Arguments](https://docs.pyinfra.com/en/3.x/arguments.html)
  * [Operation meta & callbacks](https://docs.pyinfra.com/en/3.x/arguments.html#operation-meta-callbacks): missed opportunity to discuss how the `.name` field is used.
  * [Execution strategy](https://docs.pyinfra.com/en/3.x/arguments.html#execution-strategy): Which is the default strategy, what does a value of `0` for `._parallel` mean?
# [FAQ](https://docs.pyinfra.com/en/3.x/faq.html)
  * [How do I chmod or chown ...](https://docs.pyinfra.com/en/3.x/faq.html#how-do-i-chmod-or-chown-a-file-directory-link): is that a stray all caps "LINK" in the answer?
# [Deploy Reference -> Operations Index](https://docs.pyinfra.com/en/3.x/operations.html)
  * The opening sentence sounds imperative in nature, rather than clearly either imperative or declarative.  It says it "describe[s] changes to make to systems" but if the system already meets the specified target state, no change is made.  This is the same point noted above regarding the "Using Operations" section.  It is modified by the second sentence, but overall it could be clearer.  Maybe change the opening sentence to something like "Operations execute specified commands or ensure specified target state conditions on the systems in the inventory."
  * [Files Operations](https://docs.pyinfra.com/en/3.x/operations/files.html): these are probably feature requests masquerading as doc comments, but it's surprising there is no file operation to apply patches, or set `mtime` values.  I see #877, but nothing obvious on setting `mtime`.  Are these "patches welcome" areas, or are they specifically being excluded?
# [How pyinfra works](https://docs.pyinfra.com/en/3.x/deploy-process.html)
  * In general, this whole section doesn't seem user-focused.  Subsection titles seem like misnomers, and the examples don't comprehensively illustrate what is being said.  It is hard to build a mental model of what is really going on, and when it is happening (sequencing/phasing), on both the pyinfra side and remote host side based on what is here.
  * [How pyinfra Detects Changes & Orders Operations](https://docs.pyinfra.com/en/3.x/deploy-process.html#how-pyinfra-detects-changes-orders-operations)
     * Does this section actually explain what the section title says it does?  What I get from it is that execution of operations is broken down into a series of parallel execution blocks, but no clear understanding of "why the prepare / ordering stage is so critical" beyond that.  There are no specifics as to the change detection code running on the host itself.  Maybe a more apt title would be along the lines of "How pyinfra builds a sequence of parallel [operations] execution blocks"?
     * It's not clear as-written why "if we do it in parallel we can’t ensure operations run in expected order" matters.  On a per-host basis, if you built a host-specific chain of all the operations, you could conceivably execute those per-host chains in parallel.  I believe what is meant here is that if operations on a given host are dependent on operations that would have executed on another host previously, based on how it's sequenced in the deploy, the intended sequence could be violated by running all these execution chains in parallel.  This could really benefit from a more comprehensive example and/or diagram right there.  It seems the [When does this matter?](https://docs.pyinfra.com/en/3.x/deploy-process.html#when-does-this-matter) tries to explain this, but deferring the explanation and using a non-comprehensive example instead just adds to the murkiness.
     * The penultimate sentence in the section is also confusing when it says "we have to first execute the code to generate the order" - is there any part of the deploy itself that actually gets executed beyond developing the host block sequence for execution?  This confusion is compounded when it says "This is why deploy code is always executed before any changes are made."  I _think_ what this is saying is that the act of running the deploy code creates the sequence of parallel operation execution blocks.  The way this occurs is that invoking an operation doen't immediately execute any actions on the host (other than maybe gathering facts?).  Instead, the operations return deferred execution handles.  After the whole block sequence is constructed, the sequence is walked by pyinfra and the handles are invoked to execute actions.  Is that essentially correct?  If so, I think it is less confusing if explained that way.
  * If the above is essentially correct, then how are the state checking vs. host modification steps layered in?  Is the execution of an operation "atomic", in that the state condition is checked and the host immediately modified if the state isn't the target, or is all state checking done in a batch per block, then modifications done in a second phase of that batch?  Something else?
  * There is a missed opportunity here to discuss how operations are turned into host shell code.
  * Is a host only accessed once (_e.g._, single ssh session) over the course of a whole deploy, or are there circumstances under which a sequence of connections to the host are made?
 
If you read this far, thank you.  At least for me, if the above were addressed, I'd be a lot clearer on how pyinfra works and how best to use/not use it.

**Reply by Fizzadar** [reply]
Thank you for the super detailed write up @vram0gh2! Much of this has been incorporated in the refreshed documentation in 3.9.

**Comment by bad** [comment]
It would also help if the descriptions of the api functions would match the function prototypes.
The other week I noticed files.line() describes an Append option that can't be found in the prototype.
Last summer I noticed more of these discrepancies, but there were so many that I didn't take notes as it seemed to me that a complete audit of the documentation would be necessary.

**Reply by Fizzadar** [reply]
100% agreed, and we should automate it.

**Comment by Routhinator** [comment]
Are there.. any modules anywhere for doing common things? I'm looking for better examples as the examples repo is kind of weak.

I'm trying to find things like:
- How to dearmor a GPG key now that the apt-key is deprecated and you need to put them in trusted.gpg.d after running gpg -dearmor on them.
- How to get modern APT source syntax
- How to do serialized processes on a cluster like upgrading nodes in place where only one node can be down at a time

**Reply by Routhinator** [reply]
Another couple:
* Where do facts _go_? I see we write them as classes and get how to create them, but are we expected to inline or manually import the class in every deployment file or do they get automatically registered and resolved by putting them in a specific file or location? Is there a `<module>/facts` folder in the structure? Do we put them in `group_data/all.py`?
* Where is the document covering the structure outline? I've managed to piece out the following from bits from docs, issues, the examples repo `reorg` branch `foundationdb-cluster` module, which I found because I was looking for any sign of a more complete example... so I've managed to gather this so far:

        ```
        module/
             group_data/
                  all.py
             inventories/
                  <groupname>.py
             tasks/
                  <taskname>.py     
             templates/
                  <template>.j2
             <action>.py
        ```

    But we should have some sort of outline of folder structure in the repo here.

**Reply by Routhinator** [reply]
Another - can't seem to get the SwapEnabled fact that is demonstrated to be of use.. 

```python
from pyinfra import host
from pyinfra.facts.server import Arch
from pyinfra.operations import apt, files, server, ssh
from pyinfra.api import FactBase

class SwapEnabled(FactBase):
    '''
    Returns a boolean indicating whether swap is enabled.
    '''

    command = 'swapon --show'

    def process(self, output):
        return len(output) > 0  # we have one+ lines

server.shell(
    name="Disable Swap",
    commands=["swapoff -a"],
    _if=host.get_fact(SwapEnabled),
)

```

```
--> Disconnecting from hosts...
--> Argument type error in /home/routhinator/projects/routh-forgejo/routh-ca/pyinfra/kubernetes_cluster/tasks/node_base_system_configuration.py line 16: Invalid argument `_if`:: bool did not match any element in the union:
  List[Callable[list, bool]]: is not a list
  Callable[list, bool]: is not callable
  NoneType: is not an instance of NoneType
```

The fact returns a bool, so is there some sort of hidden `is_true()` callable on facts that make it possible to use them for `_if`? This is another key thing that seems to not be covered in docs.

EDIT:

Worked out that we pass the bool through a lambda, would be good to demo that for facts docs.

```python

server.shell(
    name="Disable Swap",
    commands=["swapoff -a"],
    _if=lambda: host.get_fact(SwapEnabled),
)
```

**Reply by Routhinator** [reply]
Turns out - the Server.shell appears to fire no matter whether the SwapEnabled result is True or False :man_facepalming:

**Comment by Dexmachi** [comment]
I've been working a lot with Pyinfra recently (almost 9 months now) while building an IaC framework of my own on top of it, as I do believe It is the best IaC API out there. Due to my current experience, I felt I should write a comment here!

# Init Flow
One of my biggest issues is that, while the docs explain the individual parts relatively well, it fails on building a complete mental model on how they work together. I had to dig through the source code quite a few times to actually understand things, and even with AI it was a pain.

For instance, when using the API, the docs show a code snippet with a specific flow: inventory -> config -> state -> connect -> add_op.

## HOWEVER
The actual flow is something more like inventory -> config -> state - > state.current_stage = StateStage.Prepare -> ctx_state.set(state) -> connect -> add_op. The two parts that are important here are StateStage and ctx_state, which are not documented.

# About API interface
Another small friction point was how op results behave
```py
op = add_op(state, ...)

# To get the stdout for a specific host
op[host].stdout

# to get if all hosts changed
op.changed
```

I understand the reasoning behind op.changed representing a global change state, but the interface feels a bit inconsistent because the object is partially dictionary-like and partially object-like.

something like op["all"].changed or op.host(host).changed would be more consistent (I'd say the second one is better because of LSPs)


# About Ops and Facts
Another area that deserves significantly more attention is creating custom Facts and Operations.

Pyinfra simply does not have the same number of modules as Ansible. That's perfectly fine, but it means that the ability to create your own extensions becomes extremely important.

The project already makes this quite painless, but the documentation doesn't highlight it enough.

Examples showing how to build:

  - custom Facts

  - custom Operations

  - custom Connectors

would greatly improve discoverability and long-term maintainability of the ecosystem.


# Callback system (bro, fr)
Something similar happens with the callback system.

It's incredibly powerful but not very discoverable in the documentation.

Using callbacks I was able to build a telemetry module that captures detailed execution data, including:

  - operation timing

  - command execution

  - host-level results

This allows building quite advanced observability around Pyinfra-based tooling.

However, this capability isn't very visible in the docs today.

# Additionally (bigass request, but this could be cool)
the add_op function works and is great, I DO NOT have an issue with it. What I do have an issue with is the fact that my LSP doesn't like it (cause of static analysis).

I think that, instead of sunsetting the add_op implementation, an object could be made, something like OpAdder, that simply runs add_op's based on pyinfra objects passed onto it.
something like
```py
OpAdder.add(pacman.package(...))
```
why? Because like this we could have an type safe, easy to call and LSP compliant method to simply add ops. This is one of the main drawbacks in Pyinfra in API mode, we simply lose one of the project's main features, the LSP, because of the use of **kwargs in add_op

# summary
Overall, I’d say the main friction point for me was understanding the initialization/setup stage of the API and discovering some of the more powerful extension points.

That said, once I got past that learning curve, pyinfra has become one of my favorite IaC tools (I’m literally building my own infrastructure framework on top of it.)

**Reply by Dexmachi** [reply]
I think OpAdder could look something like so?

```py
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from typing import Callable, ParamSpec, TypeVar

    from pyinfra.api.host import Host
    from pyinfra.api.state import State

from pyinfra.api.operation import add_op

P = ParamSpec("P")
T = TypeVar("T")


class OpAdder:
    """
    Type-safe wrapper around pyinfra's add_op.
    Preserves LSP autocomplete and type checking for operations.
    """

    def __init__(self, state: State, host: Host):
        self.state = state
        self.host = host

    def add(self, op_func: Callable[P, T], *args: P.args, **kwargs: P.kwargs) -> dict:
        """
        Adds an operation to the state. The LSP will automatically
        enforce that *args and **kwargs match `op_func`.
        """

        if "host" not in kwargs:
            kwargs["host"] = self.host

        return add_op(self.state, op_func, *args, **kwargs)
```

**Reply by Fizzadar** [reply]
Thank you for the great detail, @Dexmachi! Most of this has been incorporated to the refreshed docs in 3.9!

Re: `add_op` I wonder if this could be done/similar w/the existing https://github.com/pyinfra-dev/pyinfra/blob/3.x/src/pyinfra/api/arguments_typed.py

**Comment by TyIsI** [comment]
I think API is partly unclear and under documented as well.

E.g. I had to do a lot of reading/reverse engineering to understand how to use ShortFactBase vs FactBase.

This is super annoying if (collective) you want the ecosystem to grow.

Also... Documenting that you can also use a config.py config file would be an improvement!

Elaborating on how the system effective works by using generators. (It doesn't need any default `yield` statements in `process` and it will automatically keep on looping over it until it doesn't yield anything anymore. I think.)

I'd be happy to throw some PRs in for this.

**Comment by TyIsI** [comment]
Oh! And document that developers need to update `pyinfra-metadata.toml` when adding new facts or operations.

**Reply by Routhinator** [reply]
Is there anything about this file in the docs today? Search is not bringing it up - I may be looking for just this.

**Reply by TyIsI** [reply]
@Routhinator I do not believe there is currently.

**Comment by pascal-cm** [comment]
I think a banner should be added to the previous release documentation, to give the user a heads up about potentially outdated information. 

Currently, 2.x and 3.x docs look pretty much the same. (Sometimes I end up on the old docs when searching for _pyinfra files operations_ etc.)

<img width="6144" height="3456" alt="image" src="https://github.com/user-attachments/assets/f9f9251a-a05a-49ed-8c0a-3b39959c3060" />

For reference, see Vue.js:

<img width="6144" height="3456" alt="image" src="https://github.com/user-attachments/assets/3eaeb5d4-83b4-42d3-a378-1ba6e83919f6" />

**Reply by Fizzadar** [reply]
This is a really good flag, let's fix that.

**Reply by pascal-cm** [reply]
> This is a really good flag, let's fix that.

Need a PR? I've got another two nano PRs to send in, could do in a batch.

**Reply by Fizzadar** [reply]
The new docs refresh has this (except there's a bug with the actual current docs, fixed for next release).

**Comment by Routhinator** [comment]
More feedback - this documentation needs _working_ examples. I have been trying to work out how to write a custom operation to check if swap is enabled, and I can't get a working basic operation from the documentation. 



https://docs.pyinfra.com/en/3.x/api/operations.html#example-managing-files 



This example does not work, would not yield apparently required (?) components of the return like "connector_arguments" 


```
AttributeError: 'bool' object has no attribute 'connector_arguments'
```

Thats also skipping over how the code references the `OperationError` exception which based on the imports should likely be `operations.OperationError`.


The current examples should be fleshed out from the high level examples to working code snippets. They currently expect someone new to the project to fill in gaps of information that cannot be found in the docs.
